core/include/vecmem/memory/details/sycl_custom_device_atomic_ref.hpp

```diff
@@ -0,0 +1,125 @@
+/*
+ * VecMem project, part of the ACTS project (R&D line)
+ *
+ * (c) 2022-2024 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+#pragma once
+
+// Local include(s).
+#include "vecmem/memory/details/device_address_space.hpp"
+#include "vecmem/memory/memory_order.hpp"
+
+// System include(s).
+#include <type_traits>
+
+namespace vecmem {
+namespace sycl {
+
+/// Custom implementation for atomic operations in SYCL device code
+///
+/// @tparam T Type to perform atomic operations on
+/// @tparam address The device address space to use
+///
+template <typename T,
+          device_address_space address = device_address_space::global>
+class custom_device_atomic_ref {
+
+public:
+    /// @name Type definitions
+    /// @{
+
+    /// Type managed by the object
+    typedef T value_type;
+    /// Difference between two objects
+    typedef value_type difference_type;
+    /// Pointer to the value in global memory
+    typedef value_type* pointer;
+    /// Reference to a value given by the user
+    typedef value_type& reference;
+
+    /// @}
+
+    /// @name Check(s) on the value type
+    /// @{
+
+    static_assert(
+        std::is_integral<value_type>::value,
+        "vecmem::sycl::custom_device_atomic_ref only accepts built-in "
+        "integral types");
+
+    /// @}
+
+    /// Constructor, with a pointer to the managed variable
+    explicit custom_device_atomic_ref(reference ref);
+    /// Copy constructor
+    custom_device_atomic_ref(const custom_device_atomic_ref& parent);
```

Comment:
I removed all uses of the `VECMEM_HOST_AND_DEVICE` macro from this class. As a SYCL compiler (which is the only one that should ever see this code) would never care about that macro anyway. :thinking: